auto-value-gson/src/test/java/com/ryanharter/auto/value/gson/AutoValueGsonExtensionTest.java

```diff
@@ -809,4 +812,121 @@ public void simple() {
+        + "    private String defaultB = null;\n"
+        + "    private String defaultC = null;\n"
+        + "    private String defaultD = null;\n"
+        + "    public TestTypeAdapter(Gson gson) {\n"
+        + "      this.aAdapter = gson.getAdapter(String.class);\n"
+        + "      this.bAdapter = gson.getAdapter(String.class);\n"
+        + "      this.cAdapter = gson.getAdapter(String.class);\n"
+        + "    }\n"
+        + "    public GsonTypeAdapter setDefaultA(String defaultA) {\n"
+        + "      this.defaultA = defaultA;\n"
+        + "      return this;\n"
+        + "    }\n"
+        + "    public GsonTypeAdapter setDefaultB(String defaultB) {\n"
+        + "      this.defaultB = defaultB;\n"
+        + "      return this;\n"
+        + "    }\n"
+        + "    public GsonTypeAdapter setDefaultC(String defaultC) {\n"
+        + "      this.defaultC = defaultC;\n"
+        + "      return this;\n"
+        + "    }\n"
+        + "    public GsonTypeAdapter setDefaultD(String defaultD) {\n"
+        + "      this.defaultD = defaultD;\n"
+        + "      return this;\n"
+        + "    }\n"
+        + "    @Override\n"
+        + "    public void write(JsonWriter jsonWriter, Test object) throws IOException {\n"
+        + "      if (object == null) {\n"
+        + "        jsonWriter.nullValue();\n"
+        + "        return;\n"
+        + "      }\n"
+        + "      jsonWriter.beginObject();\n"
+        + "      jsonWriter.name(\"a\");\n"
+        + "      aAdapter.write(jsonWriter, object.a());\n"
+        + "      jsonWriter.name(\"b\");\n"
+        + "      bAdapter.write(jsonWriter, object.b());\n"
+        + "      jsonWriter.endObject();\n"
+        + "    }\n"
+        + "    @Override\n"
+        + "    public Test read(JsonReader jsonReader) throws IOException {\n"
+        + "      if (jsonReader.peek() == JsonToken.NULL) {\n"
+        + "        jsonReader.nextNull();\n"
+        + "        return null;\n"
+        + "      }\n"
+        + "      jsonReader.beginObject();\n"
+        + "      String a = this.defaultA;\n"
+        + "      String b = this.defaultB;\n"
+        + "      String c = this.defaultC;\n"
+        + "      String d = this.defaultD;\n"
+        + "      while (jsonReader.hasNext()) {\n"
+        + "        String _name = jsonReader.nextName();\n"
+        + "        switch (_name) {\n"
+        + "          case \"a\": {\n"
+        + "            a = aAdapter.read(jsonReader);\n"
+        + "            break;\n"
+        + "          }\n"
+        + "          case \"c\": {\n"
+        + "            c = cAdapter.read(jsonReader);\n"
+        + "            break;\n"
+        + "          }\n"
+        + "          default: {\n"
+        + "            jsonReader.skipValue();\n"
+        + "          }\n"
+        + "        }\n"
+        + "      }\n"
+        + "      jsonReader.endObject();\n"
+        + "      return new AutoValue_Test(a, b, c, d);\n"
```

Comment:
I agree, this is good input. I'll take another look at this as soon as I have time. Thanks for the comments.